sponsor-dapp-v2/src/components/Step3.js

```diff
@@ -1,76 +1,120 @@
-import React, { Component } from "react";
+import React, { useState, useRef } from "react";
 
 import classNames from "classnames";
+import moment from "moment";
+import web3 from "web3";
 
 import IconSvgComponent from "components/common/IconSvgComponent";
+import { useIdentifierConfig } from "lib/custom-hooks";
 
-class Step3 extends Component {
-  constructor(props) {
-    super(props);
+function getContractNameAndSymbol(selections) {
+  // The date is supposed to look like Sep19 in the token name.
+  const date = moment.unix(selections.expiry).format("MMMDD");
 
-    this.state = {
-      allowedToProceed: true,
-      isLoading: false,
-      contractName: this.props.contractName,
-      editingContractName: false,
-      tokenSymbol: this.props.tokenSymbol,
-      editingTokenSymbol: false
-    };
-  }
+  // TODO(mrice32): nice to have - predict the first 4 digits of the deployed contract hash.
+  const hash = web3.utils.randomHex(2);
+
+  // Strip the "/" character from the string, so it appears as BTCUSD rather than BTC/USD.
+  const asset = selections.identifier.replace("/", "");
 
-  checkProceeding = status => {
-    this.setState({
-      allowedToProceed: status
-    });
+  const assetShort = asset.substr(0, 3);
+
+  // If either of these have been edited previously, use the previous value.
+  return {
+    name: selections.name ? selections.name : `${asset}_${date}_${hash}`,
+    symbol: selections.symbol ? selections.symbol : `${assetShort}${hash}`
   };
+}
+
+function Step3(props) {
+  const {
+    userSelectionsRef: { current: selections }
+  } = props;
+
+  const { name, symbol } = getContractNameAndSymbol(selections);
+
+  const identifierConfig = useIdentifierConfig();
+
+  const [state, setState] = useState({
+    allowedToProceed: true,
+    isLoading: false,
+    contractName: name,
+    editingContractName: false,
+    tokenSymbol: symbol,
+    editingTokenSymbol: false
+  });
 
-  handleClick(event) {
+  const contractNameTextRef = useRef(null);
+  const contractSymbolTextRef = useRef(null);
+
+  const handleClick = event => {
     event.preventDefault();
     event.persist();
 
-    this.setState(
-      {
-        isLoading: true
-      },
-      () => this.props.onNextStep(event)
-    );
-  }
+    setState(oldState => ({
+      ...oldState,
+      isLoading: true
+    }));
+
+    props.onNextStep(event);
+  };
 
-  editContractName = () => {
-    this.setState({
+  const editContractName = () => {
+    setState(oldState => ({
+      ...oldState,
       editingContractName: true
-    });
+    }));
   };
 
-  saveContractName = event => {
+  const saveContractName = event => {
     event.preventDefault();
 
-    let val = this.refs.contractNameText.value;
+    let val = contractNameTextRef.current.value;
 
-    this.setState({
+    // Only persist the name at a higher level if it was changed.
+    if (val !== state.contractName) {
```

Comment:
Can you explain why we need to avoid persisting at the higher level?

Reply:
I can add a summarized version of this as a comment.

Let's say the user comes to this page after having chosen ETH/USD as their price feed. The generated name is "ETHUSD_Sep16_0x8031".

There are two cases that I think should be treated differently:
1. User comes to step 3, doesn't change the name (or opens the field and closes it without changing it), goes back to step 1 and change the price feed to BTC/USD.
2. Same as 1, but they do change the name when they come to step 3 the first time.

Let's say the user then comes back to step 3. In 1, the name will be regenerated to be "BTCUSD_Sep16_0x4a72" because we didn't persist the default name. In 2, since the user picked a name, we keep that name.

I'm not sure this subtlety matters that much, but I defaulted to being consistent with step 1 and 2 where the fields are persisted. I'm also open to throwing out the persistence of these fields altogether so they're regenerated each time the user leaves and comes back to this page since the generation of them depends on the previous settings.